### ISSUE TITLE
Cancel a job if timeout exceed

### ISSUE THREAD

**Issue by CristianCardosoA** [Kotlin/kotlinx.coroutines#439]
I found a little problem, if the timeout is exceeded, the job is continue working until it finished.
The expected behavior that I supposed is that the job is going to be cancel if the time is exceeded. 

> waitForTheServerToSetUp() is a suspend function 

This is the code 

          val result = withTimeoutOrNull(mTimeoutDownloadDefault){

                job = launch(exceptionLogger) {

                          try {

                                download()

                          } catch (e : Exception) {

                                 mError = true
                          }
                }

                job.join()
       }

       if(result == null){
          mError = true
       }

        private suspend fun download() {

        val job = launch {

            APIClient.getLifeEventTypes(this@DownloadManager).await()
        }
        job.join()
    }

**Comment by CristianCardosoA** [comment]
The problem I think as you said is not cooperative.

```
APIClient.getLifeEventTypes 
```
Is a **suspendCancellableCoroutine**, but the timeout doesn't cancel async task.

**Comment by qwwdfsad** [comment]
Then please provide a self-contained snippet of code which reproduces the problem, I'll take a look

**Comment by qwwdfsad** [comment]
It looks like you forgot to instantiate a parent:

```
private suspend fun download() {

    val job = launch {
        APIClient.getLifeEventTypes(this@DownloadManager).await()
    }
    job.join()
}

```

When block responsible for downloading is cancelled, neither inner `launch` is cancelled nor inner `async`.

To properly cancel it, you should pass `coroutineContext` or `parent` explicitly:

```
private suspend fun download() {

    val job = launch(coroutineContext) {
        // getLifeEventTypes should start with async(CommonPool, parent = coroutineContext[Job])
        APIClient.getLifeEventTypes(this@DownloadManager).await()
    }
    job.join()
}
```

Hopefully it will be fixed automagically with #410

**Comment by CristianCardosoA** [comment]
Thanks ! I will try that and I'll post my results based on your comments.
